### ISSUE TITLE
ArgumentError: argument too big in RubyString

### ISSUE THREAD

**Issue by mjc** [jruby/jruby#3243]
Tried to generate a 10GB string in Ruby 1.7.20 and it looks like the number is too big for RubyString?

```
irb(main):001:0> GB = 1024*1024*1024; a = "a" * GB * 10
ArgumentError: argument too big
	from org/jruby/RubyString.java:1195:in `*'
	from (irb):1:in `evaluate'
	from org/jruby/RubyKernel.java:1111:in `eval'
	from org/jruby/RubyKernel.java:1511:in `loop'
	from org/jruby/RubyKernel.java:1274:in `catch'
	from org/jruby/RubyKernel.java:1274:in `catch'
```

MRI successfully generates this string.

**Comment by enebo** [comment]
I was going to change int as index to long as index in RubyString but I realize that we will run into platform issues (which is likely why this was originally capped at MAX_INT).  Without significant changes to how RubyString is defined (like using ropes/multiple lists internally) this will not get solved easily.

I am leaving this open as I think we can solve this but I do not anticipate us fixing this near term.

**Comment by mjc** [comment]
Is there another way to generate a massive string you'd recommend in the interim?;

**Comment by enebo** [comment]
@mjc Perhaps make a string-like class which aggregates a list of strings.  This may not sound appealing but it does have a useful property:  Not all memory needs to be contiguously allocated.  Finding a single contiguous region of 10G of memory is potentially challenging but 10 1G segments much less so.

**Comment by rtyler** [comment]
@mjc out of curiosity, have you tried creating very large strings in MRI as well? I wonder if we're behaving appropriately here

**Comment by mjc** [comment]
Yes, this works fine in MRI, as big a string as memory and swap allow.

**Comment by enebo** [comment]
@rtyler MRI is long so they may work or likely work for larger than max int

**Comment by headius** [comment]
It's not possible for us to represent a large string like this with standard JVM code, since a byte[] in Java can only be indexed by a 32-bit signed integer. It would be possible to do this if we had a ByteList that used tricks like Unsafe, but that's a really big leap to make.

**Comment by enebo** [comment]
If we make aggregate bytelist with @nirvdrum bytelist extension changes we can support it through that but then we would need long signatures on bytelist (jotting notes here)
